subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c

```diff
@@ -84,17 +91,86 @@ enum {
 	/* Request recieved */
 	RP_PU_EVT_PHY_REQ,
 
+	/* Ack received */
+	RP_PU_EVT_ACK,
+
 	/* Indication recieved */
 	RP_PU_EVT_PHY_UPDATE_IND,
 };
 
-/*
- * LLCP Local Procedure PHY Update FSM
- */
+/* Hardcoded instant delta +6 */
+#define PHY_UPDATE_INSTANT_DELTA  6
+#define GENERATE_NTF  		1U
+#define DONT_GENERATE_NTF 	0U
+
+static inline uint8_t select_phy(uint8_t _x)
+{
+	/* select only one phy, prefer 2M */
+	if (_x & PHY_2M) {
+		return PHY_2M;
+	} else if (_x & PHY_1M) {
+		return PHY_1M;
+	} else if (_x & PHY_CODED) {
+		return PHY_CODED;
+	} else {
+		return 0U;
+	}
+}
+
+static void util_prep_update_ind(struct ll_conn *conn, struct proc_ctx *ctx)
+{
+
+	ctx->data.pu.tx = select_phy(ctx->data.pu.tx);
+	ctx->data.pu.rx = select_phy(ctx->data.pu.rx);
+
+	/* Initiate PHY Update Ind */
+	if (ctx->data.pu.tx != conn->lll.phy_tx) {
+		ctx->data.pu.m_to_s_phy = ctx->data.pu.tx;
+	} else {
+		ctx->data.pu.m_to_s_phy = 0U;
+	}
+	if (ctx->data.pu.rx != conn->lll.phy_rx) {
+		ctx->data.pu.s_to_m_phy = ctx->data.pu.rx;
+	} else {
+		ctx->data.pu.s_to_m_phy = 0U;
+	}
+}
+
+static void util_apply_timing_restrict(struct ll_conn *conn, uint8_t phy_tx)
+{
+	switch (conn->lll.role) {
+	case BT_HCI_ROLE_MASTER:
+		conn->lll.phy_tx_time = phy_tx;
+		break;
+	case BT_HCI_ROLE_SLAVE:
+		{
+			/* select the probable PHY with longest Tx time, which
+			 * will be restricted to fit current
+			 * connEffectiveMaxTxTime.
+			 */
+			uint8_t phy_tx_time[8] = {PHY_1M, PHY_1M, PHY_2M,
```

Comment:
No, I guess not - I inherited this code snippet. I guess it could be optimized a tiny bit, however, it would only make it less readable in my opinion.